### DISCUSSION TITLE
Best way to reuse same connection or pool connections in NextJs

### DISCUSSION THREAD

**Discussion by 1finedev** [Automattic/mongoose#12249] [General]
i was wondering if the code below is the best way to avoid connecting to mongoDb every time an api request is made in nextJs

I am hoping anyone who has used mongoDb and nextJs in production can advise me so i don't run into the "too many connections" error 

```js
import mongoose from "mongoose";

const { MONGODB_URI, MONGODB_DB } = process.env;

if (!MONGODB_URI) {
  throw new Error(
    "Please define the MONGODB_URI environment variable inside .env.local"
  );
}

if (!MONGODB_DB) {
  throw new Error(
    "Please define the MONGODB_DB environment variable inside .env.local"
  );
}

export async function connectToDatabase() {
  const opts = {
    useNewUrlParser: true,
    useUnifiedTopology: true,
    autoIndex: true,
  };

  let client;
  let clientPromise;

  if (!process.env.MONGODB_URI) {
    throw new Error("Please add your Mongo URI to .env.local");
  }

  if (process.env.NODE_ENV === "development") {
    // In development mode, use a global variable so that the value
    // is preserved across module reloads caused by HMR (Hot Module Replacement).
    if (!global._mongoClientPromise) {
      global._mongoClientPromise = mongoose
        .connect(MONGODB_URI, opts)
        .then((client) => {
          return {
            client,
          };
        });
    }
    clientPromise = global._mongoClientPromise;
    console.log("db connected successfully");
  } else {
    // In production mode, it's best to not use a global variable.
    clientPromise = mongoose.connect(MONGODB_URI, opts).then((client) => {
      return {
        client,
      };
    });
    console.log("db connected successfully");
  }

  // Export a module-scoped MongoClient promise. By doing this in a
  // separate module, the client can be shared across functions.
  return clientPromise;
}
```

**Comment by vkarpov15** [comment]
Take a look at our Lambda docs: https://mongoosejs.com/docs/lambda.html

**Reply by 1finedev** [reply]
thanks @vkarpov15 but NextJs does not expose a context to be able to use on this line ```context.callbackWaitsForEmptyEventLoop = false;```, or am i missing something?

**Reply by vkarpov15** [reply]
`context.callbackWaitsForEmptyEventLoop` isn't necessary with Next, I think they do that by default. Here's NextJS' example for Mongoose: https://github.com/vercel/next.js/tree/canary/examples/with-mongodb-mongoose

**Comment by wedothebest4you** [comment]
The following posts may help.

Dealing with a database with Next.js API routes
https://www.kevincunningham.co.uk/posts/dealing-with-a-database-with-nextjs-api-routes/

How do you share a database connection between API routes in NextJS 13?
https://stackoverflow.com/questions/77024733/how-do-you-share-a-database-connection-between-api-routes-in-nextjs-13

What Happens When a Module Is Imported Twice?
https://dmitripavlutin.com/javascript-module-import-twice/
